numpy/core/src/multiarray/datetime.c

```diff
@@ -1761,6 +1762,9 @@ parse_datetime_unit_from_string(char *str, Py_ssize_t len, char *metastr)
                 return NPY_FR_as;
         }
     }
+    else if (allow_generic && len == 7 && !strncmp(str, "generic", 7)) {
```

Comment:
Pickling non-generic datetime dtypes works, so the fact that pickling generic datetime dtypes fails really fails like a bug rather than a deliberate decision, no? `descriptor.c` has a dedicated `_get_pickleabletype_from_datetime_metadata` function after all, it's just that it didn't work in all cases.